File: rob599_hw3/src/timer.py
#!/usr/bin/env python3

import sys
import rospy

from sensor_msgs.msg   import LaserScan


class timer:

	def __init__(self, cycles=100):

		self.sub1 = rospy.Subscriber('base_scan',    LaserScan, self.callback1, queue_size=1   )
		self.sub2 = rospy.Subscriber('timer/clocked_scan', LaserScan, self.callback2,  queue_size=100 )
		self.pub  = rospy.Publisher( 'timer/queued_scan',  LaserScan, queue_size=100 )

		self.manifest = {}
		self.count = 0
		self.length = cycles
		self.times = []




	def callback1(self, lidar_msg):
		"""
		Callback function to filter LaserScan Data.
		"""
		if self.count == self.length:
			self.count = self.count+1

			ave = sum(self.times)/len(self.times)

			rospy.loginfo("Timming is complete, Ave: {}".format(ave))

			f = open("/home/ubuntu/rob599_homeWork/py_average_time.txt","w")
			f.write( "Average Time: {}".format(ave) )
			f.close()

			rospy.loginfo("Ave saved to file")

			return

		elif self.count > self.length:
			return

		self.count = self.count+1

		self.manifest[self.count] = rospy.get_time()

		self.pub.publish(lidar_msg)



	def callback2(self, lidar_msg):

		dt = rospy.get_time()  - self.manifest[lidar_msg.header.seq]

		self.times.append(dt)

		rospy.loginfo("Time: {0}".format(dt))



if __name__ == '__main__':
	# Initialize the node.
	rospy.init_node('laser_timmer')

	t = timer()

	rospy.loginfo("'laser_timmer' Node Started\n")

	# Set up a subscriber and publisher.

	rospy.spin()

File: rob599_hw3/src/timer.cpp
/*
	Jacob Anderson
	ROB599 HW3
	Dec 10, 2020
*/

#include <vector>
#include <ros/ros.h>
#include "sensor_msgs/LaserScan.h"
#include <numeric>
#include <fstream>

class timer_node {


protected:
	ros::NodeHandle node;                 //  The Node Handel
	ros::Subscriber sub1;                 //  Subscribe to the base scan as our source of laser scans
  ros::Subscriber sub2;                 //  Subscribe to the returning scan lines
	ros::Publisher  pub;                  //  Publish the scan lines that are being timed

  std::vector<double> trip_times;       //  Duration for the other nodes to do their work
  std::map<int, ros::Time> manifest;    //  Keeps track of when the outgoing scan was sent
  int count;


public:

	timer_node(){

		// Set up the subscriber and publisher
    sub1 = node.subscribe("base_scan", 1, &timer_node::callback1, this);
    sub2 = node.subscribe("timer/clocked_scan", 100, &timer_node::callback2, this);
		pub = node.advertise<sensor_msgs::LaserScan>("timer/queued_scan", 1);

    // Initialize the count
    count = 0;
	}


	void callback1(const sensor_msgs::LaserScan::ConstPtr& scan){             	//  Callback to send out the laser scans

		if(count > 100) return;                                                 	//  Stop after 100 Laser Scans

    count++;                                                                	//  Keep track of how many laser scans have been sent

    manifest.insert(std::pair<int, ros::Time> (count, ros::Time::now()) );  	//  Make note of current time

		pub.publish(*scan);                                                     	//  Send out the scan line
	}


  void callback2(const sensor_msgs::LaserScan::ConstPtr& scan){             	// Callback for the returning laser scans

  	if ( manifest.find(scan->header.seq) == manifest.end() ) return;          //  Make Sure the key exists

  	ros::Duration dt = ros::Time::now()- manifest[scan->header.seq];          //  Elapsed time for nodes to work

  	trip_times.push_back(dt.toSec());                                         //  Hold onto the duration

  	ROS_INFO("Duration: %f", dt.toSec() );

  	if(scan->header.seq == 100){                                              //  After 100 messages, find the average time

    	double average = std::accumulate( trip_times.begin(), trip_times.end(), 0.0) / trip_times.size();

    	ROS_INFO("Average: %f\n\n\n\n", average);

			// Write Avarage to file
			std::ofstream myfile;
  		myfile.open ("/home/ubuntu/rob599_homeWork/cpp_average_time.txt");
  		myfile << "Average Time:" << average << std::endl;
  		myfile.close();

    	}
  }

};




int main(int argc, char **argv) {
	// Initialize the node and set up the node handle.
	ros::init(argc, argv, "timer_noiser");

	timer_node tn;

  ROS_INFO("Timer Node Started");
	// Give control over to ROS.
	ros::spin();

	return 0;
}
